File: 2-D Dynamic Programming/3202. Find the Maximum Length of Valid Subsequence II/solution.py
class Solution:
    def maximumLength(self, nums: List[int], k: int) -> int:
        n = len(nums)
        dp = [[0]*k for _ in range(k)]

        res = 1
        for num in nums:
            x = num%k
            for m in range(k):
                prev = (m-num)%k
                dp[x][m] = max(dp[x][m], dp[prev][m]+1)
                res = max(res, dp[x][m])
        return res

File: 2-D Dynamic Programming/3202. Find the Maximum Length of Valid Subsequence II/solution.cpp
class Solution {
public:
    int maximumLength(vector<int>& nums, int k) {
        int n = nums.size();
        for (int i=0; i<n; i++) {
            nums[i] %= k;
        }

        vector<vector<int>> dp(k, vector<int>(k, 0));
        int res = 1;
        for (auto& num: nums) {
            for (int m=0; m<k; m++) {
                int prev = ((m-num)+k)%k;
                dp[num][m] = max(dp[num][m], dp[prev][m]+1);
                res = max(res, dp[num][m]);
            }
        }
        return res;
    }
};